lib/model.js

```diff
@@ -2683,11 +2683,22 @@ class Model {
         // Map updateOnDuplicate attributes to fields
         if (options.updateOnDuplicate) {
           options.updateOnDuplicate = options.updateOnDuplicate.map(attr => model.rawAttributes[attr].field || attr);
-          // Get primary keys for postgres to enable updateOnDuplicate
-          options.upsertKeys = _.chain(model.primaryKeys).values().map('field').value();
-          if (Object.keys(model.uniqueKeys).length > 0) {
-            options.upsertKeys = _.chain(model.uniqueKeys).values().filter(c => c.fields.length >= 1).map(c => c.fields).reduce(c => c[0]).value();
+
+          const upsertKeys = [];
+
+          for (const i of model._indexes) {
+            if (i.unique) {
```

Comment:
probably worth making sure that the index isn't a partial index - if its a partial index then it won't apply unless that where clause is specified, which currently we don't do.

Reply:
Done at 9bc372a